server/hotrod/src/main/java/org/infinispan/server/hotrod/ClientListenerRegistry.java

```diff
@@ -330,10 +338,16 @@ boolean hasChannel(Channel channel) {
 
       // This method can only be invoked from the Event Loop thread!
       void writeEventsIfPossible() {
+         boolean submittedUnblock = false;
          boolean written = false;
          while (!eventQueue.isEmpty() && ch.isWritable()) {
+            int newSize = eventSize.decrementAndGet();
             Events.Event event = eventQueue.poll();
             if (isTrace) log.tracef("Write event: %s to channel %s", event, ch);
+            if (!submittedUnblock && newSize < lowWaterMark && !delayedEvents.isEmpty()) {
+               addListenerExecutor.execute(this::unblockCommands);
+               submittedUnblock = true;
```

Comment:
If you could write to the channel more than `(highWaterMark - lowWaterMark) * 2` events without blocking, you could in theory start writing, unblock once, have a burst of events that reach the high water mark again and block, then write everything from the event queue without unblocking.

I'm not sure if that's possible, but it would be better IMO to either unblock only once at the end (simpler) or unblock once we hit the low water mark + again at the end if we're still below the low water mark.